asset/asset.go

```diff
@@ -524,19 +524,32 @@ func NewRawKeyGenesisSigner(priv *btcec.PrivateKey) *RawKeyGenesisSigner {
 // identified by the passed key descriptor. The final tweaked public key and
 // the signature are returned.
 func (r *RawKeyGenesisSigner) SignGenesis(keyDesc keychain.KeyDescriptor,
-	gen Genesis) (*btcec.PublicKey, *schnorr.Signature, error) {
+	groupGen Genesis, newGen *Genesis) (*btcec.PublicKey,
+	*schnorr.Signature, error) {
 
 	if !keyDesc.PubKey.IsEqual(r.privKey.PubKey()) {
 		return nil, nil, fmt.Errorf("cannot sign with key")
 	}
 
 	tweakedPrivKey := txscript.TweakTaprootPrivKey(
-		*r.privKey, gen.GroupKeyTweak(),
+		*r.privKey, groupGen.GroupKeyTweak(),
 	)
 
+	// The key tweak and signed genesis are the same when first using
+	// a group key. For future use, the asset type of the new asset must
+	// match the first asset of the group.
+	if newGen == nil {
+		newGen = &groupGen
+	} else if groupGen.Type != newGen.Type {
+		return nil, nil, fmt.Errorf(
```

Comment:
Seems like we're missing signing and verification here. Perhaps we can just move the verification to a later stage: when we double back over to verify that we're signing the asset genesis properly.

Reply:
Independent of this comment: commit body could use a bit more detail w.r.t rationale of this change.

Reply:
Yup, I think rn this typecheck is performed when fetching the input `Genesis` structs / before ever calling `DeriveGroupKey`.